File: revision/unique_char.cpp
#include<iostream>
#include<vector>
#include<iostream>
#include<vector>
#include<unordered_map>
using namespace std;
void display(vector<int>nums){
    for(auto const i : nums){
        cout << i << ' ' ;
    }
    cout << endl;
}
using namespace std;
int uniqueChar(string &s){
    //using fixed 26 char array 
    // vector<int>count(26,0);
    // for(char i : s){
    //     count[int(i) - int('a')]+=1;
    // }
    // for(int i = 0; i < s.size(); i++){
    //     if (count[int(s[i]) - int('a')] == 1){
    //         return i;
    //     }
    // }
    // return -1;
    
    //using hashmap 
    unordered_map<char,int>myhashmap;
    for(char c : s){
        if (myhashmap.find(c) == myhashmap.end()){
            myhashmap.insert(make_pair(c, 1));
        } else{
            myhashmap[c]++;
        }
    }
    for(int i = 0; i < s.size(); i++){
        if(myhashmap.at(s[i]) == 1){
            return i;
        }
    }
    return -1;
}
int main()
{
    string s;
    cin >> s;
    cout << uniqueChar(s)<<endl;
    return 0;
}

File: revision/unique_char.py
from typing import List
def firstUniqueChar(s:str)->int:
    '''
    Using HASHMAP
    s = "leetcode"
    l: 1
    e: 3
    t: 1
    c: 1
    o: 1
    d: 1
    '''
    # mydict = dict()
    # for x in s:
    #     mydict[x] = mydict.get(x, 0) + 1
    # for i, v in enumerate(s):
    #     if mydict[v] == 1:
    #         return i
    # return -1
    '''
    using fixed array
    s = "leetcode"
    count = bucket of 26 letters: [0 0 1 0 2 0 0 0 0 0 0 1 0 ] 
                                   a b c d e f g h i j k l       
    iterate through the string
    for i in range(len(nums)):
        c = ord(nums[i]) - ord('a')
        if count[c] == 1:
            return i
    return -1
    '''
    count = [0] * 26
    for x in s:
        count[ord(x) - ord('a')]+=1
    print(count)
    for i in range(len(s)):
        if count[ord(s[i]) - ord('a')] == 1:
            return i
    return -1      
    
print(firstUniqueChar(input()))
